### PR TITLE
Refactor (frontend): typed pages

### PR DESCRIPTION
**Pull request by luixo**
Step 2 of [getting complete typescript coverage](https://github.com/near/near-explorer/issues/819)

### CODE REVIEW

File: frontend/src/components/accounts/AccountDetails.tsx [hunk 1]
```diff
@@ -173,74 +173,75 @@ class AccountDetails extends Component<Props> {
...
-                          currentNearNetwork.nearWalletProfilePrefix
-                        }
-                      />
-                    }
-                  />
-                </Col>
-              </Row>
-            )}
+            {typeof account.nonStakedBalance !== "undefined" &&
+              typeof account.stakedBalance !== "undefined" && (
```
Comment (luixo) [inline frontend/src/components/accounts/AccountDetails.tsx:177]: Apparently they only come together, but we need to comply with types

File: frontend/src/components/utils/Content.tsx [hunk 2]
```diff
@@ -10,8 +10,7 @@ export interface Props {
   fluid?: boolean;
   contentFluid?: boolean;
   icon?: React.ReactNode;
-  className: string;
-  children: React.ReactNode;
```
Comment (luixo) [inline frontend/src/components/utils/Content.tsx:14]: Or should I keep it to mark that children are mandatory?

File: frontend/src/context/DatabaseProvider.tsx [hunk 3]
```diff
@@ -25,11 +31,7 @@ export interface DbContext {
 
 const DatabaseContext = createContext<DbContext>({});
 
-export interface Props {
-  children: React.Component;
```
Comment (luixo) [inline frontend/src/context/DatabaseProvider.tsx:29]: The same as previous comment

File: frontend/src/context/NodeProvider.tsx [hunk 4]
```diff
@@ -12,11 +12,7 @@ export interface INodeContext {
 
 const NodeContext = createContext<INodeContext>({});
 
-export interface Props {
-  children: React.Component;
```
Comment (luixo) [inline frontend/src/context/NodeProvider.tsx:16]: And one more mandatory `children`

File: frontend/src/libraries/language.ts [hunk 5]
```diff
@@ -27,27 +42,20 @@ function normalizeLocales(arr) {
...
+  let languageList: string[] = [];
 
   if (typeof window !== "undefined") {
     const { navigator } = window;
 
     if (navigator.languages) {
       languageList = languageList.concat(navigator.languages);
     }
+    // https://caniuse.com/mdn-api_navigator_language
+    // 97.31% users have this property as of 2021-12-15
```
Comment (luixo) [inline frontend/src/libraries/language.ts:55]: Is that good enough for us? Do we have a profile of our users?

File: frontend/src/libraries/language.ts [hunk 6]
```diff
@@ -88,24 +96,30 @@ export function getBrowserLocale(appLanguages) {
...
-export function getAcceptedLocale(appLanguages, acceptedLanguages) {
-  acceptedLanguages =
+export function getAcceptedLocale(
+  appLanguages: string[],
+  acceptedLanguages?: string
+): string | undefined {
+  const parsedAcceptedLanguages =
     acceptedLanguages && parseAcceptLanguage(acceptedLanguages);
 
   debugLog("Languages from Accepted Languages header:", acceptedLanguages);
```
Comment (luixo) [inline frontend/src/libraries/language.ts:106]: Woops, should put `parsedAcceptedLanguages` here as well

File: frontend/src/libraries/language.ts [hunk 7]
```diff
@@ -138,13 +152,14 @@ function findBestSupportedLocale(appLocales, browserLocales) {
         const existingMatch = matchedLocales[matchedPartialLocale];
 
         // Deduct a thousandth for being non-exact match.
+        const newMatchScore = 0.999 - index / walletLocales.length;
```
Comment (luixo) [inline frontend/src/libraries/language.ts:155]: That's a bug that has been discovered with types.
It still exists in a wallet, we should probably report it after merging that.

File: frontend/src/libraries/language.ts [hunk 8]
```diff
@@ -242,11 +266,10 @@ export function getI18nConfigForProvider({ cookies, acceptedLanguages }) {
   }
 }
 
-export function setI18N(props) {
+export function setI18N(props: LocalizeContextProps): void {
   const config = getI18nConfig();
   const { languages } = config;
-  const { cookies } = props;
```
Comment (luixo) [inline frontend/src/libraries/language.ts:248]: There are actually no `cookies` in props, I need a sanity check here

File: frontend/src/pages/_app.tsx [hunk 9]
```diff
@@ -35,15 +41,18 @@ class _App extends App {
...
+        currentNearNetwork = getNearNetwork(req.headers.host);
+        cookies = req.headers.cookie;
+        acceptedLanguages = req.headers["accept-language"];
+      }
     } else {
       currentNearNetwork = getNearNetwork(window.location.host);
     }
     return {
       currentNearNetwork,
-      ...(await App.getInitialProps({ ...appContext, currentNearNetwork })),
```
Comment (luixo) [inline frontend/src/pages/_app.tsx:46]: Why did [we put](https://github.com/near/near-explorer/commit/f6e22c0b41ec1852b04b5e8165d501a2873049f2#diff-c74b166bc48fd32a6900a92b516a36c250f268be81d57463ce342172c3148ecbL25-R39) `currentNearNetwork` here? It seems like `App.getInitialProps` doesn't do anything with it and nothing seems broken after I removed it.

File: frontend/src/pages/accounts/[id].tsx [hunk 10]
```diff
@@ -14,10 +15,27 @@ import Content from "../../components/utils/Content";
...
+  accountError?: unknown;
+}
+
+class AccountDetail extends Component<Props> {
+  static async getInitialProps({
+    req,
+    query: { id: rawId },
+    res,
+  }: NextPageContext) {
+    // Change to string[] if pagename changes from `[id]` to `[...id]`
```
Comment (luixo) [inline frontend/src/pages/accounts/[id].tsx:36]: It this comment a little too obvious?

File: frontend/src/pages/stats/index.tsx [hunk 11]
```diff
@@ -51,7 +51,7 @@ class Stats extends PureComponent {
           </div>
           <NetworkConsumer>
             {({ currentNearNetwork }) =>
-              currentNearNetwork.name === "mainnet" ? (
+              currentNearNetwork?.name === "mainnet" ? (
```
Comment (luixo) [inline frontend/src/pages/stats/index.tsx:54]: Those `?`s seem a little excess. We probably should make a hook which encapsulate network existence or make a default network

File: frontend/src/libraries/language.ts (hunk 7 above)
Comment (shelegdmitriy) [reply frontend/src/libraries/language.ts:155]: @luixo you can create an issue on [wallet](https://github.com/near/near-wallet)

File: frontend/src/pages/_app.tsx (hunk 9 above)
Comment (frol) [reply frontend/src/pages/_app.tsx:46]: I don’t see the reason either. I must have refactored something along the way and it was a leftover from some intermediate steps. Good catch!

File: frontend/src/components/utils/Content.tsx (hunk 2 above)
Comment (shelegdmitriy) [reply frontend/src/components/utils/Content.tsx:14]: From my point of view `children` is required everywhere exept of  `frontend/src/pages/contracts/index.jsx` but I can't find any place where we use it. @luixo can we set `FC` to `children`? With `className?: string` - agree

File: frontend/src/pages/accounts/[id].tsx (hunk 10 above)
Comment (frol) [reply frontend/src/pages/accounts/[id].tsx:36]: Well, it is a bit hard to reason about this comment at first glance, but after a second thought it is obvious and next I thought of “why did you tell me this?”. We either need to rephrase it or just drop it (there is no meaningful use-case for […id])

File: frontend/src/pages/stats/index.tsx (hunk 11 above)
Comment (frol) [reply frontend/src/pages/stats/index.tsx:54]: My only concern would be that we may potentially hide some bugs lurking due to the fallback always hiding non-set value, but it seems it should not be the case with the current network

**Review by frol — APPROVED** [review]
👍

**Review by shelegdmitriy — APPROVED** [review]

File: frontend/src/pages/accounts/[id].tsx (hunk 10 above)
Comment (luixo) [reply frontend/src/pages/accounts/[id].tsx:36]: Let's remove it, actually

File: frontend/src/pages/stats/index.tsx (hunk 11 above)
Comment (luixo) [reply frontend/src/pages/stats/index.tsx:54]: I'll make encapsulate it in a hook in a PR soon, it'll throw error if there is no current network.

File: frontend/src/components/utils/Content.tsx (hunk 2 above)
Comment (luixo) [reply frontend/src/components/utils/Content.tsx:14]: @shelegdmitriy
Sorry, I didn't get you
I removed three excess `children: React.ReactNode` in `Context.tsx`, `DatabaseProvider.tsx` and `NodeProvider.tsx`, how is contracts page connected to that?
All three cases seem to render completely fine without children

Comment (shelegdmitriy) [reply frontend/src/components/utils/Content.tsx:14]: @luixo when I was searching for places where we use `Content` I realized that we always pass children to this component exept for [this one](https://github.com/near/near-explorer/blob/5e9807af2093c2aa85ba8ee596f860d87e30115c/frontend/src/pages/contracts/index.tsx#L12). But the thing is thet we doesn't use this page anywhere. And so we can remove this file as well. 
But nevermind, no action here

Comment (luixo) [reply frontend/src/components/utils/Content.tsx:14]: Щл

File: frontend/Dockerfile [hunk 12]
```diff
@@ -20,7 +20,6 @@ ENV HOME=/tmp \
 COPY --from=build /near-explorer/production_node_modules ./node_modules
 COPY --from=build /near-explorer/frontend/public ./public
 COPY --from=build /near-explorer/frontend/next.config.js ./
-COPY --from=build /near-explorer/frontend/src/libraries/consts.js ./src/libraries/consts.js
```
Comment (luixo) [inline frontend/Dockerfile:23]: @frol 
I removed `consts.js` from here, it seems like it's pretty much unneeded now

File: frontend/next.config.ts [hunk 13]
```diff
@@ -10,6 +9,10 @@ export interface NearNetwork {
   nearWalletProfilePrefix: string;
 }
 
+enum DataSource {
+  LegacySyncBackend = "LEGACY_SYNC_BACKEND",
```
Comment (luixo) [inline frontend/next.config.ts:13]: It's the only data source still used in client code

**Pull request merged**